cli,sdnotify: improve the handling of unix socket directories

knz opened the pull request:
tldr: This patch makes it easier for users to understand when
CockroachDB has a problem using/creating unix sockets, and
what they can do to avert the situation.

Fixes #76904.

Example, before;
```
$ cockroach start-single-node --insecure --background
...
listen unixgram /data/home/kena/cockroach/this-is-a-very-long-directory-name-the-point-is-to-be-more-than-one-hundred-and-twenty-three-characters/sdnotify2139543236/notify.sock: bind: invalid argument
```

Example, after:
```
$ cockroach start-single-node --insecure --background
ERROR: no suitable directory found for the --background notify socket
HINT: Avoid using --background altogether (preferred), or use a
shorter directory name for --socket-dir, TMPDIR or the current directory.
```

**Context:**

CockroachDB uses unix sockets both to handle `--background` and to
create a postgres-compatible SQL client socket via `--socket-dir`. For
`--background`, the logic was previously hardwired to always use
`os.TempDir()`, i.e. either `$TMPDIR` or `/tmp`.

On Unix, socket objects are commonly limited to 128 bytes;
that's e.g. 104 path characters including final NUL on BSD, 108 on glibc.
When the actual length is larger, creating the socket fails with
a confusing error `bind: invalid argument`.

**The problem:**

Users have reported issues in the following circumstances:

- when setting their CWD/$TMPDIR to a long directory name,
  they expected `--socket-dir` to also influence the socket
  name used for `--background`. Instead, `--background`
  failed with a hard-to-understand error. (#76904)

- when running tests inside a container/sandbox with a long
  prefix, they were confused that `--background` / `--socket-dir=.`
  did not work properly.

- they use macOS which has very large directory names in $TMPDIR.

To improve user experience, this patch changes the behavior
as described in the release notes below below.

Release note (cli change): CockroachDB now produces a clearer error
when the path specified via `--socket-dir` is too long.

Release note (cli change): When the flag `--background` is specified,
CockroachDB now makes 3 attempts to find a suitable directory to
create the notification socket: the value of `--socket-dir` if
specified, the value of `$TMPDIR` (or /tmp if the env var is empty),
and the current working directory. If none of these directories has a
name short enough, an explanatory error is printed.

knz committed:
cli,sdnotify: improve the handling of unix socket directories

tldr: This patch makes it easier for users to understand when
CockroachDB has a problem using/creating unix sockets, and
what they can do to avert the situation.

Example, before;
```
$ cockroach start-single-node --insecure --background
...
listen unixgram /data/home/kena/cockroach/this-is-a-very-long-directory-name-the-point-is-to-be-more-than-one-hundred-and-twenty-three-characters/sdnotify2139543236/notify.sock: bind: invalid argument
```

Example, after:
```
$ cockroach start-single-node --insecure --background
ERROR: no suitable directory found for the --background notify socket
HINT: Avoid using --background altogether (preferred), or use a
shorter directory name for --socket-dir, TMPDIR or the current directory.
```

**Context:**

CockroachDB uses unix sockets both to handle `--background` and to
create a postgres-compatible SQL client socket via `--socket-dir`. For
`--background`, the logic was previously hardwired to always use
`os.TempDir()`, i.e. either `$TMPDIR` or `/tmp`.

On Unix, socket objects are commonly limited to 128 bytes;
that's e.g. 104 path characters including final NUL on BSD, 108 on glibc.
When the actual length is larger, creating the socket fails with
a confusing error `bind: invalid argument`.

**The problem:**

Users have reported issues in the following circumstances:

- when setting their CWD/$TMPDIR to a long directory name,
  they expected `--socket-dir` to also influence the socket
  name used for `--background`. Instead, `--background`
  failed with a hard-to-understand error.

- when running tests inside a container/sandbox with a long
  prefix, they were confused that `--background` / `--socket-dir=.`
  did not work properly.

- they use macOS which has very large directory names in $TMPDIR.

To improve user experience, this patch changes the behavior
as described in the release notes below below.

Release note (cli change): CockroachDB now produces a clearer error
when the path specified via `--socket-dir` is too long.

Release note (cli change): When the flag `--background` is specified,
CockroachDB now makes 3 attempts to find a suitable directory to
create the notification socket: the value of `--socket-dir` if
specified, the value of `$TMPDIR` (or /tmp if the env var is empty),
and the current working directory. If none of these directories has a
name short enough, an explanatory error is printed.

knz committed:
cli/start: remove noise when shutting down after `--backround`

Release note (bug fix): The server process does not any more announce
that it is shutting down on stdout when running with `--background`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84532)
<!-- Reviewable:end -->

knz committed:
pgwire,sdnotify: shortcut to /tmp in tests when necessary

This commit changes `TestAuthenticationAndHBARules` (pgwire) and
`TestSDNotify` (sdnotify) to use `/tmp` as socket directory if the
default temporary directory name is too long.

Release note: None

jreut reviewed (approved):
LGTM. I see mention of long macOS $TMP paths. Have these tests simply historically failed on macOS?

> Reviewed 7 of 7 files at r1, 1 of 1 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84532#-:-N7MkujF5trtx_Qf90T1:bxalr8t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @stevendanna)

___
*[`pkg/cli/flags.go` line 1013 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84532#-N7MjTFVAQ52afydbgKa:-N7MjTFVAQ52afydbgKb:b8ycu70) ([raw file](https://github.com/cockroachdb/cockroach/blob/265667cab6fa3c98548d58f4f87521e4d53b7541/pkg/cli/flags.go#L1013)):*
> ```Go
> 		} else {
> 			socketName := ".s.PGSQL." + serverListenPort
> 			// On BSD, binding to a socket is limited to a path length of 104 characters
> ```

For my own education, could you link to the code or docs that support these facts?
___
*[`pkg/cli/interactive_tests/test_socket_name.tcl` line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84532#-N7Mk4sKDU3D3eqzKnG4:-N7Mk4sKDU3D3eqzKnG5:bd6njic) ([raw file](https://github.com/cockroachdb/cockroach/blob/265667cab6fa3c98548d58f4f87521e4d53b7541/pkg/cli/interactive_tests/test_socket_name.tcl#L65)):*
> ```Tcl
> system "$crdb sql --insecure -e 'select 1'"
> stop_server $crdb
> end_test
> ```

TIL we have TCL-based tests in CRDB. These are a joy to read!


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Have these tests simply historically failed on macOS?

They do when you try to run them outside of the docker sandbox.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84532#-:-N7Mman364nrMwAf3mT3:b1g3qvx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @stevendanna)

___
*[`pkg/cli/flags.go` line 1013 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84532#-N7MjTFVAQ52afydbgKa:-N7Mm2QE4y9pmNVSCqgE:bco51qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/265667cab6fa3c98548d58f4f87521e4d53b7541/pkg/cli/flags.go#L1013)):*
<details><summary><i>Previously, jreut (jordan ryan reuter) wrote…</i></summary><blockquote>

For my own education, could you link to the code or docs that support these facts?
</blockquote></details>

https://www.man7.org/linux/man-pages/man7/unix.7.html
```
      A UNIX domain socket address is represented in the following
       structure:

           struct sockaddr_un {
               sa_family_t sun_family;               /* AF_UNIX */
               char        sun_path[108];            /* Pathname */
           };

       The sun_family field always contains AF_UNIX.  On Linux, sun_path
       is 108 bytes in size
```

https://www.freebsd.org/cgi/man.cgi?query=unix&sektion=4
```
UNIX-domain addresses are variable-length file system pathnames of	at
     most 104 characters.  The include file <sys/un.h> defines this address:

	   struct sockaddr_un {
		   u_char  sun_len;
		   u_char  sun_family;
		   char	   sun_path[104];
	   };
```


<!-- Sent from Reviewable.io -->

knz commented:
TFYR!

bors r=jreut
